FAST AI Lesson - 4

Importing all the important libraries

In [1]:
%load_ext autoreload
%autoreload 2

This will help us in plotting using matplotlib

In [2]:
%matplotlib inline

Importing Fastai libraries

In [3]:
from fastai.nlp import *

Importing other libraries

In [4]:
from sklearn.linear_model import LogisticRegression

For NLP(Natural Language Processing) we are going to work on the IMDB movie view dataset.

# IMDB Dataset

Path to the dataset and then we are printing out the files in /train/pos directory.

In [5]:
path = 'data/aclImdb/'
names = ['neg', 'pos']

In [6]:
%ls {path}

imdbEr.txt  imdb.vocab  README  test/  train/


In [7]:
%ls {path}train

labeledBow.feat  pos/    unsupBow.feat  urls_pos.txt
neg/             unsup/  urls_neg.txt   urls_unsup.txt


In [8]:
%ls {path}train/pos | head

0_9.txt
10000_8.txt
10001_10.txt
10002_7.txt
10003_8.txt
10004_8.txt
10005_7.txt
10006_7.txt
10007_7.txt
10008_7.txt
ls: write error


texts_labels_from_folders() function basically to the directory f'{path}train' and then in that directory go through
each file start with words in names list. And then divide it in training and test set and allot the value to it based 
on names list. Like for the data from neg folder will be given the value 0 and then the data from pos will be given 1.

In [9]:
trn, trn_y = texts_labels_from_folders(f'{path}train', names)
val, val_y = texts_labels_from_folders(f'{path}test', names)

In [10]:
trn[0]

'I\'m gettin\' sick of movies that sound entertaining in a one-line synopsis then end up being equal to what you\'d find in the bottom center of a compost heap.<br /><br />Who knows: "Witchery" may have sounded interesting in a pitch to the studios, even with a "big name cast" (like Blair and Hasselhoff - wink-wink, nudge-nudge) and the effervescent likes of Hildegard Knef (I dunno, some woman...).<br /><br />But on film, it just falls apart faster than a papier-mache sculpture in a rainstorm. Seems these unfortunate folks are trapped in an island mansion off the Eastern seaboard, and one of them (a woman, I\'d guess) is being targeted by a satanic cult to bear the child of hell while the others are offed in grotesque, tortuous ways. <br /><br />Okay, right there you have a cross-section of plots from "The Exorcist", "The Omen", "Ten Little Indians" and a few other lesser movies in the satanic-worshippers-run-amok line. None of it is very entertaining and for the most part, you\'ll cri

Label for the above text is 0. Means negative because we have neg on 0th index in names list.

In [11]:
trn_y[0]

0

In [12]:
len(trn)

25000

So our job is to take trn and predict trn_y for each text data in testing set.
For this we are going to create a document matrix and then we will remove the order of word for this. This Unordered 
text here is called bag of words which tell us about the words that are in that document.
So document matrix is matrix of token counts. For this we are going to do following.

In [13]:
vec = CountVectorizer(tokenizer=tokenize)

In [14]:
vec

CountVectorizer(analyzer='word', binary=False, decode_error='strict',
        dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
        lowercase=True, max_df=1.0, max_features=None, min_df=1,
        ngram_range=(1, 1), preprocessor=None, stop_words=None,
        strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
        tokenizer=<function tokenize at 0x7faf7b80bea0>, vocabulary=None)

In [15]:
trn_doc = vec.fit_transform(trn)
val_doc = vec.transform(val)

So now we have created a matrix for based on the training data for this we make the use of fit_transform()
and then we are transform() function just to create a matrix based on validation data but we are taking the same order
of word from the training data document matrix.

In [16]:
trn_doc

<25000x75132 sparse matrix of type '<class 'numpy.int64'>'
	with 3749745 stored elements in Compressed Sparse Row format>

As we can see that the dimension for the trn_doc is 25000 which is number of text documents in training set and 
75132 is the number of different words that we came across these 25000 documents

In [17]:
val_doc

<25000x75132 sparse matrix of type '<class 'numpy.int64'>'
	with 3640465 stored elements in Compressed Sparse Row format>

Now lets grab the first row which for the 1st movie review

In [18]:
trn_doc[0]

<1x75132 sparse matrix of type '<class 'numpy.int64'>'
	with 244 stored elements in Compressed Sparse Row format>

So 244 words in total are used from 75132. Now we are grabbing the some of the words from the 1st text documents.

In [19]:
vocab = vec.get_feature_names()

In [20]:
vocab[5000:50005]

['aussie',
 'aussies',
 'austen',
 'austeniana',
 'austens',
 'auster',
 'austere',
 'austerity',
 'austin',
 'austinese',
 'austion',
 'austrailan',
 'austrailia',
 'austrailian',
 'australia',
 'australian',
 'australians',
 'australlian',
 'austreheim',
 'austreim',
 'austria',
 'austrialian',
 'austrian',
 'austrians',
 'austro',
 'autant',
 'auteil',
 'auteuil',
 'auteur',
 'auteurist',
 'auteurs',
 'authenic',
 'authentic',
 'authentically',
 'authenticating',
 'authenticity',
 'author',
 'authored',
 'authoress',
 'authorial',
 'authoring',
 'authorisation',
 'authorised',
 'authoritarian',
 'authoritarianism',
 'authoritative',
 'authoritatively',
 'authorities',
 'authority',
 'authorize',
 'authorized',
 'authorizes',
 'authorizing',
 'authors',
 'authorship',
 'authorty',
 'authur',
 'autie',
 'autism',
 'autistic',
 'auto',
 'autobiograhical',
 'autobiographic',
 'autobiographical',
 'autobiography',
 'autobots',
 'autocockers',
 'autocracy',
 'autocratic',
 'autocue',
 'au

So now we are splitting the text document 1 from the training set based on space (' ')and taking all the lowercase 
words and their length is almost same as the one calculated in document matrix.

In [21]:
w = set([o.lower() for o in trn[0].split(' ')])
w

{'"all".',
 '"big',
 '"mark',
 '"ten',
 '"the',
 '"witchery"',
 '(a',
 '(and',
 '(i',
 '(like',
 '-',
 '/><br',
 '/>and',
 '/>but',
 "/>don't",
 '/>from',
 '/>no',
 '/>no,',
 '/>okay,',
 '/>who',
 '<br',
 'a',
 'an',
 'and',
 'anderson',
 'apart',
 'are',
 'as',
 'back',
 'band.<br',
 'beack',
 'bear',
 'being',
 'believe',
 'benefit',
 'blair',
 'blame',
 'bother.',
 'bottom',
 'bursting',
 'by',
 'can',
 "can't",
 'cast"',
 'celebrity',
 'center',
 'child',
 'cliches.',
 'coming',
 'compost',
 'count',
 'cringe',
 'cross-section',
 'cult',
 'dangled',
 'david',
 'devil"',
 'did.',
 "didn't",
 'do',
 "don't",
 'down',
 'dunno,',
 'during',
 'eastern',
 'ecchhhh.',
 'effervescent',
 'end',
 'entertaining',
 'equal',
 'etc,',
 'etc.',
 'even',
 'exorcist",',
 'falls',
 'faster',
 'few',
 'film,',
 'find',
 'fireplaces,',
 'folks',
 'followed',
 'for',
 'from',
 'get',
 'gets',
 "gettin'",
 'give',
 'giving',
 'golf',
 'grotesque,',
 'guess)',
 'hasselhoff',
 'have',
 'he',
 'heap.<br',


In [22]:
len(w)

243

In [23]:
vec.vocabulary_['absurd']

1297

Above function will give the column number of the word in document matrix. Document matrix doesn't care about the 
relative ordering of the word.

# Naive Bayes

Log count ratio r for each word f is log(ratio of that word in +ve doc/ratio of that word in -ve doc)

In [24]:
x = trn_doc
x

<25000x75132 sparse matrix of type '<class 'numpy.int64'>'
	with 3749745 stored elements in Compressed Sparse Row format>

In [25]:
y = trn_y
y

array([0, 0, 0, ..., 1, 1, 1])

Now we have grabs all the rows with y==1 and then sum it and plus 1. Similarly for the y==0.

In [26]:
p = x[y==1].sum(0)+1
q = x[y==0].sum(0)+1

In [27]:
p

matrix([[    2,     1, 11820, ...,     2,     2,     1]], dtype=int64)

In [28]:
p.shape

(1, 75132)

In [29]:
q

matrix([[    1,     2, 12742, ...,     1,     2,     8]], dtype=int64)

In [30]:
q.shape

(1, 75132)

In [31]:
r = np.log((p/p.sum())/(q/q.sum()))

In [32]:
r

matrix([[ 0.68213, -0.70417, -0.08613, ...,  0.68213, -0.01102, -2.09046]])

In [33]:
r.shape

(1, 75132)

So r is the log count ratio.

In [34]:
b = np.log(len(p)/len(q))

In [35]:
b

0.0

Applying Naive Bayes Formula

In [36]:
pre_preds = val_doc @ r.T + b
preds = pre_preds.T>0
(preds==val_y).mean()

0.8074

Applying the binarized version of Naive Bayes. In this .sign() finction will convert the anything >0 with 1 and <0 with
-1.

In [37]:
pre_preds = val_doc.sign() @ r.T + b
preds = pre_preds.T > 0
(preds==val_y).mean()

0.82624

# Logistic Regression

In [38]:
x

<25000x75132 sparse matrix of type '<class 'numpy.int64'>'
	with 3749745 stored elements in Compressed Sparse Row format>

In [39]:
y

array([0, 0, 0, ..., 1, 1, 1])

Dual parameter will make the Logistic Regression run faster on Larger Dataset.

In [40]:
m = LogisticRegression(C=1e8, dual=True)
m.fit(x, y)
preds = m.predict(val_doc)
(preds==val_y).mean()

/home/priyanshu/anaconda3/envs/fastai/lib/python3.6/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/home/priyanshu/anaconda3/envs/fastai/lib/python3.6/site-packages/sklearn/svm/base.py:931: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


0.85768

For the Binarized version of Naive Bayes

In [41]:
m = LogisticRegression(C=1e8, dual=True)
m.fit(trn_doc.sign(), y)
preds = m.predict(val_doc.sign())
(preds==val_y).mean()

/home/priyanshu/anaconda3/envs/fastai/lib/python3.6/site-packages/sklearn/svm/base.py:931: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


0.85488

Since we are dealing with huge number of parameter, so there is a chance that our model can overfit. So we have to 
regularized it. We can make the use of C parameter in order to regularized the data. More C value, less regularization

In [42]:
m = LogisticRegression(C=0.1, dual=True)
m.fit(trn_doc, y)
preds = m.predict(val_doc)
(preds==val_y).mean()

/home/priyanshu/anaconda3/envs/fastai/lib/python3.6/site-packages/sklearn/svm/base.py:931: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


0.88272

In [43]:
m = LogisticRegression(C=0.1, dual=True)
m.fit(trn_doc.sign(), y)
preds = m.predict(val_doc.sign())
(preds==val_y).mean()

0.88404

Cross Entropy is another loss function used for classification purpose. 
Formula is CE = act_y*log(pre_y)-(1-act_y)*log(1-pre_y), where act_y = actual classification and pre_y = predicted 
classificatiom

For classification purpose use Cross Entropy as a loss function and for Regression use RMSE loss function

# Trigram with NB Feature

Now we are going to add more features by adding digram and trigram to our dataset. So we are taking only top 800000 features into consideration and then adding the features to the document matrix.

In [44]:
vec = CountVectorizer(ngram_range=(1,3), tokenizer=tokenize, max_features=800000)

In [45]:
trn_doc = vec.fit_transform(trn)
val_doc = vec.transform(val)

In [46]:
trn_doc.shape

(25000, 800000)

Getting some of the features from 800000 features.

In [48]:
vocab = vec.get_feature_names()

In [50]:
vocab[200000:200005]

['by vast', 'by vengeance', 'by vengeance .', 'by vera', 'by vera miles']

In [51]:
y = trn_y
y

array([0, 0, 0, ..., 1, 1, 1])

In [52]:
val_y

array([0, 0, 0, ..., 1, 1, 1])

Converting all the value > 0 to 1 and all the value < 0 to -1

In [54]:
x = trn_doc.sign()
x

<25000x800000 sparse matrix of type '<class 'numpy.int64'>'
	with 12589101 stored elements in Compressed Sparse Row format>

In [55]:
val_x = val_doc.sign()
val_x

<25000x800000 sparse matrix of type '<class 'numpy.int64'>'
	with 11555871 stored elements in Compressed Sparse Row format>

Now applying Naive Bayes to it

In [56]:
p = x[y==1].sum(0)+1
q = x[y==0].sum(0)+1
r = np.log((p/p.sum())/(q/q.sum()))
b = np.log(len(p)/len(q))

In [57]:
pre_preds = val_x @ r.T + b
preds = pre_preds.T>0
(preds==val_y).mean()

0.88044

Now applying logistic regression on it.

In [58]:
m = LogisticRegression(C=0.1, dual=True)
m.fit(x,y)
preds = m.predict(val_x)
(preds==val_y).mean()

/home/priyanshu/anaconda3/envs/fastai/lib/python3.6/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


0.905

In [59]:
r.shape

(1, 800000)

In [60]:
r

matrix([[-0.04911, -0.15543, -0.24226, ...,  1.10419, -0.68757, -0.68757]])

In [61]:
np.exp(r)

matrix([[0.95208, 0.85605, 0.78485, ..., 3.01678, 0.5028 , 0.5028 ]])

r is actually the log value of ratios so to get the actual ratio we need to do e^(r)

In [62]:
x_nb = x.multiply(r)
m = LogisticRegression(C=0.1, dual=True)
m.fit(x_nb,y)

val_x_nb = val_x.multiply(r)
preds = m.predict(val_x_nb)
(preds==val_y).mean()

/home/priyanshu/anaconda3/envs/fastai/lib/python3.6/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


0.91768

So why we are getting better accuracy on mulitpling the document metrix with r?? This is because initially for the 
negative document we have 0 and for positive we have 1 but after multipling it by r our value for negative document is
0.3333 and for positive we have 3. So indirectly we have increase the amount of variance between the positive and 
negative document.

# Fastai NBSVM++

In [77]:
sl = 2000

This is amount of unique we have to use for the further processing.

In [78]:
md = TextClassifierData.from_bow(trn_doc, trn_y, val_doc, val_y, sl)

Here we are getting the model from bag of words. Here md is data object which wraps both training and validation data
and this bag of words will have only 2000 unique words for model deployment.

In [79]:
learn = md.dotprod_nb_learner()

dotprod_nd_learner() function is NBSVM for the fastai library. Now we are going to fit the model

In [80]:
learn.fit(0.02, 1, wds=1e-6, cycle_len=1)

epoch      trn_loss   val_loss   <lambda>                     
    0      0.025744   0.11924    0.91664   



[array([0.11924]), 0.9166400000190735]

In [81]:
learn.fit(0.02, 1, wds=1e-6, cycle_len=1)

epoch      trn_loss   val_loss   <lambda>                     
    0      0.021512   0.113841   0.91992   



[array([0.11384]), 0.9199199999809266]

In [82]:
learn.fit(0.02, 1, wds=1e-6, cycle_len=1)

epoch      trn_loss   val_loss   <lambda>                     
    0      0.018895   0.111163   0.92208   



[array([0.11116]), 0.9220800000190735]

In [83]:
learn.fit(0.02, 1, wds=1e-6, cycle_len=1)

epoch      trn_loss   val_loss   <lambda>                     
    0      0.017486   0.110443   0.92172   



[array([0.11044]), 0.9217199999809265]

In [84]:
learn.fit(0.02, 1, wds=1e-6, cycle_len=1)

epoch      trn_loss   val_loss   <lambda>                     
    0      0.016829   0.109381   0.92244   



[array([0.10938]), 0.9224399999809265]

So after 5 epoch we are getting the accuracy of around 92.2% which is quite better then previous versions.

So in dotprod_nb_learner() function we add some constant to the wx this is beacuse regularization tries to make 
everything zero. And if everything is zero then complete matrix multiplication will be zero. So to avoid this condition
we usually add some constant most probably 0.4 to the wx, so that wx doesn't reduce to 0.